flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultBatchJobRecoveryHandler.java

```diff
@@ -0,0 +1,840 @@
+                                            Execution.createResultPartitionDeploymentDescriptor(
+                                                    partition, shuffleDescriptor));
+
+                            availablePartitionBytes.put(
+                                    shuffleDescriptor.getResultPartitionID().getPartitionId(),
+                                    partitionAndMetric.getPartitionMetrics().getPartitionBytes());
+                        });
+
+        // recover the produced partitions for executions
+        Map<
+                        ExecutionVertexID,
+                        Map<IntermediateResultPartitionID, ResultPartitionDeploymentDescriptor>>
+                allDescriptors = new HashMap<>();
+        ((InternalExecutionGraphAccessor) executionGraph)
+                .getPartitionTracker()
+                .getAllTrackedNonClusterPartitions()
+                .forEach(
+                        descriptor -> {
+                            ExecutionVertexID vertexId =
+                                    descriptor
+                                            .getShuffleDescriptor()
+                                            .getResultPartitionID()
+                                            .getProducerId()
+                                            .getExecutionVertexId();
+                            if (!allDescriptors.containsKey(vertexId)) {
+                                allDescriptors.put(vertexId, new HashMap<>());
+                            }
+
+                            allDescriptors
+                                    .get(vertexId)
+                                    .put(descriptor.getPartitionId(), descriptor);
+                        });
+
+        allDescriptors.forEach(
+                (vertexId, descriptors) ->
+                        getExecutionVertex(vertexId)
+                                .getCurrentExecutionAttempt()
+                                .recoverProducedPartitions(descriptors));
+
+        // recover result partition bytes
+        updateResultPartitionBytesMetricsFunction.accept(availablePartitionBytes);
+
+        // restart all producers of missing partitions
+        List<ExecutionVertexID> missingPartitionVertices =
+                reconcileResult.missingPartitions.stream()
+                        .map(ResultPartitionID::getPartitionId)
+                        .map(this::getProducer)
+                        .map(ExecutionVertex::getID)
+                        .collect(Collectors.toList());
+
+        // get all vertices need to restart according failover strategy.
+        Set<ExecutionVertexID> verticesToReset = new HashSet<>();
+        for (ExecutionVertexID executionVertexId : missingPartitionVertices) {
+            if (!verticesToReset.contains(executionVertexId)) {
+                verticesToReset.addAll(
+                        failoverStrategy.getTasksNeedingRestart(executionVertexId, null));
```

Comment:
Should we do this in all cases that invokes `resetVerticesInRecovering()`?

Reply:
Because after restoring the operator coordinator, at that time the upstream partition is not available. If we call `failoverStrategy.getTasksNeedingRestart`, all upstream tasks will be mistakenly included. So I do this only after partition recovered finished.